lib/services/plugin_data.go

```diff
@@ -0,0 +1,313 @@
+	data := PluginDataV3{
+		Kind:    KindPluginData,
+		Version: V3,
+		// If additional resource kinds become supported, make
+		// this a parameter.
+		SubKind: resourceKind,
+		Metadata: Metadata{
+			Name: resourceName,
+		},
+		Spec: PluginDataSpecV3{
+			Entries: make(map[string]*PluginDataEntry),
+		},
+	}
+	if err := data.CheckAndSetDefaults(); err != nil {
+		return nil, err
+	}
+	return &data, nil
+}
+
+func (d *PluginDataV3) Entries() map[string]*PluginDataEntry {
+	if d.Spec.Entries == nil {
+		d.Spec.Entries = make(map[string]*PluginDataEntry)
+	}
+	return d.Spec.Entries
+}
+
+func (d *PluginDataV3) Update(params PluginDataUpdateParams) error {
+	// See #3286 for a complete discussion of the design constraints at play here.
+
+	if params.Kind != d.GetSubKind() {
+		return trace.BadParameter("resource kind mismatch in update params")
+	}
+
+	if params.Resource != d.GetName() {
+		return trace.BadParameter("resource name mismatch in update params")
+	}
+
+	// If expectations were given, ensure that they are met before continuing
+	if params.Expect != nil {
+		if err := d.checkExpectations(params.Plugin, params.Expect); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	// Ensure that Entries has been initialized
+	if d.Spec.Entries == nil {
+		d.Spec.Entries = make(map[string]*PluginDataEntry, 1)
+	}
+	// Ensure that the specific Plugin has been initialized
+	if d.Spec.Entries[params.Plugin] == nil {
+		d.Spec.Entries[params.Plugin] = &PluginDataEntry{
+			Data: make(map[string]string, len(params.Set)),
+		}
+	}
+	entry := d.Spec.Entries[params.Plugin]
+	for key, val := range params.Set {
+		// Keys which are explicitly set to the empty string are
+		// treated as DELETE operations.
+		if val == "" {
+			delete(entry.Data, key)
+			continue
+		}
+		entry.Data[key] = val
+	}
+	// Its possible that this update was simply clearing all data;
+	// if that is the case, remove the entry.
+	if len(entry.Data) == 0 {
+		delete(d.Spec.Entries, params.Plugin)
+	}
+	return nil
+}
+
+// checkExpectations verifies that the data for `plugin` matches the expected
+// state described by `expect`.  This function implements the behavior of the
+// `PluginDataUpdateParams.Expect` mapping.
+func (d *PluginDataV3) checkExpectations(plugin string, expect map[string]string) error {
+	var entry *PluginDataEntry
+	if d.Spec.Entries != nil {
+		entry = d.Spec.Entries[plugin]
+	}
+	if entry == nil {
+		// If no entry currently exists, then the only expectation that can
+		// match is one which only specifies fields which shouldn't exist.
+		for key, val := range expect {
+			if val != "" {
+				return trace.CompareFailed("expectations not met for field %q", key)
+			}
+		}
+		return nil
+	}
+	for key, val := range expect {
+		if entry.Data[key] != val {
+			return trace.CompareFailed("expectations not met for field %q", key)
+
+		}
+	}
+	return nil
+}
+
+func (f *PluginDataFilter) Match(data PluginData) bool {
+	if f.Kind != "" && f.Kind != data.GetSubKind() {
```

Comment:
SubKind seems like a good candidate for plugin name?

Reply:
`Subkind` is already used to encode the associated resource kind.  Currently it will always be `KindAccessRequest`, but using `SubKind` in this way ensures that we won't have any conflicts if we choose to expand the scope of `PluginData` later.

Even if `SubKind` were unused, its possible for more than one plugin to store data within a singe `PluginData` resource (hence the double-layer mapping in `PluginDataSpec`).